### PR TITLE
[Discounts] Add default demo functions for new discounts APIs

### PR DESCRIPTION
**Pull request by dphm**
Rust examples for:

- `product_discounts`: percentage off products with excluded variant ids
- `order_discounts`: percentage off order with excluded variant ids
- `shipping_discounts`: percentage off shipping

🎩 

### Login to Shopify CLI

```sh
SPIN=1 shopify login
```

Use `partners@shopify.com` with Italic Apps 

### Build and push scripts

```sh
shopify script create --api=product_discounts --language=rust --branch=discounts/new-rust --title=rust_p && cd rust_p && make && SPIN=1 shopify script push
```

```sh
shopify script create --api=order_discounts --language=rust --branch=discounts/new-rust --title=rust_o && cd rust_o && make && SPIN=1 shopify script push
```

```sh
shopify script create --api=shipping_discounts --language=rust --branch=discounts/new-rust --title=rust_s && cd rust_s && make && SPIN=1 shopify script push
```

```sh
Are you working on a Shopify project on behalf of the Shopify partners org? (Choose with ↑ ↓ ⏎)
> 1. no

Which partner organization do you want to use? (Choose with ↑ ↓ ⏎, filter with 'f')
> 1. Italic Apps (1)
```

### CODE REVIEW

File: discounts/rust/order-discounts/default/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
+use serde::{Deserialize, Serialize};
+use serde_with::skip_serializing_none;
```
Comment (dphm) [inline discounts/rust/order-discounts/default/src/main.rs:2]: TODO: Figure out how to use this

File: discounts/rust/product-discounts/default/src/api.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#[serde(rename_all(serialize = "SCREAMING_SNAKE_CASE"))]
+pub enum DiscountApplicationStrategy {
+    First,
+    Maximum,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+pub struct Input {
+    pub merchandise_lines: Option<Vec<MerchandiseLine>>,
```
[the remark this answers is not in the document]
Comment (dphm) [reply discounts/rust/product-discounts/default/src/api.rs:79]: Yeah, I agree we should :P I just tried to make this work for now

File: discounts/rust/product-discounts/default/src/api.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all(serialize = "camelCase"))]
+pub enum Value {
+    FixedAmount(FixedAmount),
+    Percentage(Percentage),
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct FixedAmount {
+    pub value: MoneySubunits,
```
[the remark this answers is not in the document]
Comment (dphm) [reply discounts/rust/product-discounts/default/src/api.rs:32]: Oh yeah I forgot about that 😆 I should make an example with fixed amount too

File: discounts/rust/product-discounts/default/src/api.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        target_type: ConditionTargetType,
+        excluded_variant_ids: Vec<ID>,
+        minimum_amount: MoneySubunits,
+    },
+    #[serde(rename_all(serialize = "camelCase"))]
+    ProductMinimumSubtotal {
+        target_type: ConditionTargetType,
+        excluded_variant_ids: Vec<ID>,
+        minimum_amount: MoneySubunits,
+    },
```
[the remark this answers is not in the document]
Comment (dphm) [reply discounts/rust/product-discounts/default/src/api.rs:60]: Lol yep thanks!

**Comment by ifyouseewendy** [comment]
Update

* Manually add input types to `src/api.rs` under the `input` module
* Fix and update types for three APIs
* Tophat and get product_discount, order_discount default script working. shipping_discount script is failing with the following error (won't have time to fix today, @dphm you could help pick it up tomorrow). 

Screen Shot 2022-05-19 at 9 12 20 PM

The failure is caused by the empty `targets`. Not sure how it got to that code path.

```
May 19 20:42:53 spin shopify--shopify.server[3238798]: {
May 19 20:42:53 spin shopify--shopify.server[3238798]:   "input": {
May 19 20:42:53 spin shopify--shopify.server[3238798]:     "deliveryLines": [
May 19 20:42:53 spin shopify--shopify.server[3238798]:     ]
May 19 20:42:53 spin shopify--shopify.server[3238798]:   },
May 19 20:42:53 spin shopify--shopify.server[3238798]:   "configuration": {
May 19 20:42:53 spin shopify--shopify.server[3238798]:     "value": "33.0"
May 19 20:42:53 spin shopify--shopify.server[3238798]:   }
May 19 20:42:53 spin shopify--shopify.server[3238798]: }
May 19 20:42:53 spin shopify--shopify.server[3238798]: [ServiceProxy::Proxy] [SCRIPTS_V2] script output:
May 19 20:42:53 spin shopify--shopify.server[3238798]: {
May 19 20:42:53 spin shopify--shopify.server[3238798]:   "discountApplicationStrategy": "FIRST",
May 19 20:42:53 spin shopify--shopify.server[3238798]:   "discounts": [
May 19 20:42:53 spin shopify--shopify.server[3238798]:     {
May 19 20:42:53 spin shopify--shopify.server[3238798]:       "value": {
May 19 20:42:53 spin shopify--shopify.server[3238798]:         "percentage": {
May 19 20:42:53 spin shopify--shopify.server[3238798]:           "value": 33.0
May 19 20:42:53 spin shopify--shopify.server[3238798]:         }
May 19 20:42:53 spin shopify--shopify.server[3238798]:       },
May 19 20:42:53 spin shopify--shopify.server[3238798]:       "targets": [
May 19 20:42:53 spin shopify--shopify.server[3238798]:       ],
May 19 20:42:53 spin shopify--shopify.server[3238798]:       "message": "33% off"
May 19 20:42:53 spin shopify--shopify.server[3238798]:     }
May 19 20:42:53 spin shopify--shopify.server[3238798]:   ]
May 19 20:42:53 spin shopify--shopify.server[3238798]: }
```


One thing I've noticed that is we need a quick review over the input types. For any required field, the type will be defined as non-null. However, that field might not be included in the input query, which ends up a runtime error. I believe this issue is considered when people define the input schema in Core, as most of the fields are optional. We should consider updating all fields to optional, or at least the ones that can be reached from the top-level, eg. `price` field in `MerchandiseLine`. https://github.com/Shopify/shopify/blob/main/components/pricing/app/models/scripts_graph/pricing/discount_types_api.rb#L112-L119

**Comment by dphm** [comment]
I'll try to make that error less terrible 😆

**Comment by dphm** [comment]
I just fixed these scripts so they return empty lists of discounts instead of discounts with no targets.

**Pull request merged**